Revive beam information, physics declared bit, and HLT event selection for online DQM 81X

diguida opened the pull request:
The PR allows the DQM module `DQMProvInfo` (run by central online DQM) to retrieve the beam information from TCDS FED raw data, by reading the corresponding address for the Beam Synchronous Timing (BST) record. This is currently done using an utility in `EventFilter/Utilities`, which has been enhanced for this scope.
From this information, the Physics Declared bit is computed, and added into the report summary profile.
NOTA BENE: this is a **temporary** solution, until a well implemented unpacker for TCDS is made available in CMSSW.

The PR also resurrects a `MonitorElement` in `DQMEventInfo` showing the HLT path selection in the DQM input source `DQMStreamerReader`.

Commit:
Accessing relevant beam information from BST record in EventFilter/Utilities/plugins/testGlobalNumbers.cc.

Commit:
Add a test configuration for GlobalNumbersAnalysis EDAnalyzer reading FED data from RAW data tier.

Commit:
Use a more complex definiton of TCDSBST class in EventFilter/FEDInterface/interface/FED1024.h in order to avoid issues with padding.

Commit:
Add back the LHC information in DQMProvInfo. They are retrieved from the FED1024 raw data using the utilities in EventFilter/Utilities package.

Commit:
Revive a MonitorElement in DQMEventInfo showing the HLT path selection in the DQM input source DQMStreamerReader.

cmsbuild commented:
A new Pull Request was created by  @diguida (Salvatore Di Guida) for CMSSW_8_1_X.

It involves the following packages:

DQM/Integration
DQMServices/Components
EventFilter/FEDInterface
EventFilter/Utilities


@cvuosalo, @emeschi, @dmitrijus, @cmsbuild, @slava77, @mommsen, @vanbesien, @davidlange6 can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @barvic, @threus, @batinkov this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

diguida commented:
@jhegeman @smorovic this is something you might want to watch as well. Your support was appreciated.

slava77 commented:
@cmsbuild please test
not sure if the DAS is back to life

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: be209e425f8a38dd874c8aa2cba877d42587a346
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14763/13360/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14763/13360/summary.html

dmitrijus commented:
'temporary' as in we don't merge it?

diguida commented:
@dmitrijus 'temporary' --> this is the best we can offer at the moment for retrieving such information. The code should be re-organised, but the data sources are correct.

dmitrijus commented:
@diguida Does not answer the question, which was: should this be merged? :)

diguida commented:
@dmitrijus sorry I was not clear.
Yes, it should go in.

mommsen commented:
@emeschi this more in your expertise. Can you please look at the proposed changes? Remi

emeschi commented:
I looked at it - I think the modifications to the FEDInterface are minor - essentially renaming a few fields, and changing the test code to print more (and mostly changing indents…). Per se these modifications are not necessary for the purpose of the DQM, and I would have preferred that we take this on in a more general way, but have no strong objections

E.

> On 08 Jun 2016, at 12:17, Remi Mommsen <notifications@github.com> wrote:
> 
> @emeschi <https://github.com/emeschi> this more in your expertise. Can you please look at the proposed changes? Remi
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/cms-sw/cmssw/pull/14763#issuecomment-224548169>, or mute the thread <https://github.com/notifications/unsubscribe/AFFwXC-ZpqVMAehoH1_9t7cq5KWW9gyQks5qJpaggaJpZM4Itm1j>.
>

cvuosalo commented:
`EventFilter/Utilities/plugins/testGlobalNumbers.cc` appears to be a test program. Is there a good reason why it is in the `plugins` directory rather than the `test` directory?

cvuosalo commented on EventFilter/Utilities/plugins/testGlobalNumbers.cc:
```diff
@@ -22,177 +24,191 @@
-  private:
-  public:
-    GlobalNumbersAnalysis(const ParameterSet& pset){
+  class GlobalNumbersAnalysis: public edm::EDAnalyzer {
```
If this file is not moved to the `test` directory, the inheritance from EDAnalyzer should be changed to `edm::one::EDAnalyzer<>` to quiet the static analyzer warning.

cvuosalo commented on EventFilter/Utilities/plugins/testGlobalNumbers.cc:
```diff
@@ -22,177 +24,191 @@
-      Handle<FEDRawDataCollection> rawdata;
-      e.getByLabel("source",rawdata);
+    void analyze(const edm::Event & e, const edm::EventSetup& c) {
+      std::cout << "--- Run: " << e.id().run()
```
If this file doesn't move to the `test` directory, it would be good to change the `cout` statements to `LogDebug()`, either now or in a later PR.

emeschi replied:
+1
No reason not to move the thing to test.
Let’s take this PR and then move it.

> On 08 Jun 2016, at 22:37, Carl Vuosalo <notifications@github.com> wrote:
> 
> In EventFilter/Utilities/plugins/testGlobalNumbers.cc <https://github.com/cms-sw/cmssw/pull/14763#discussion_r66334809>:
> 
> >
> > -    void analyze(const Event & e, const EventSetup& c){
> > -      cout << "--- Run: " << e.id().run()
> > -	   << " LS: " << e.luminosityBlock()
> > -	   << " Event: " << e.id().event()
> > -	   << " Type: " << e.experimentType() << endl;
> > -      Handle<FEDRawDataCollection> rawdata;
> > -      e.getByLabel("source",rawdata);
> > +    void analyze(const edm::Event & e, const edm::EventSetup& c) {
> > +      std::cout << "--- Run: " << e.id().run()
> If this file doesn't move to the test directory, it would be good to change the cout statements to LogDebug(), either now or in a later PR.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/cms-sw/cmssw/pull/14763/files/be209e425f8a38dd874c8aa2cba877d42587a346#r66334809>, or mute the thread <https://github.com/notifications/unsubscribe/AFFwXBbEEzFPyLPpFxNtNAWnPHGUKMgLks5qJygGgaJpZM4Itm1j>.
>

dmitrijus commented:
+1

dmitrijus commented:
If I understand correctly, we include this PR and do 'move to test' with another?

emeschi commented:
Yes

Sent from my iPhone

On 9 Jun 2016, at 14:42, Dmitrijus <notifications@github.com<mailto:notifications@github.com>> wrote:


If I understand correctly, we include this PR and do 'move to test' with another?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/cms-sw/cmssw/pull/14763#issuecomment-224882903>, or mute the thread<https://github.com/notifications/unsubscribe/AFFwXCTmVlX_EnJ8czBD7bMy7NjxP__Sks5qKAo_gaJpZM4Itm1j>.

cvuosalo commented:
+1

For #14763 be209e425f8a38dd874c8aa2cba877d42587a346

Saving beam info for online DQM. There should be no change in monitored offline Reco quantities.

The code changes are satisfactory, and Jenkins tests against baseline CMSSW_8_1_X_2016-06-02-1100 show no significant differences, as expected.

Commit:
Move EventFilter/Utilities/plugins/testGlobalNumbers.cc to EventFilter/Utilities/test/testGlobalNumbers.cc, and add a BuildFile in EventFilter/Utilities/test.

Commit:
Make analyzer in EventFilter/Utilities/test/testGlobalNumbers.cc inherit from edm::one::EDAnalyzer<>.

Commit:
Use MessageLogger API in EventFilter/Utilities/test/testGlobalNumbers.cc, and further improve logging.

diguida commented:
@dmitrijus @emeschi @cvuosalo thanks for your review, and sorry for belated reply.
I have already prepared the commits addressing comments in https://github.com/cms-sw/cmssw/pull/14763#issuecomment-224717419 https://github.com/cms-sw/cmssw/pull/14763#discussion_r66334307 https://github.com/cms-sw/cmssw/pull/14763#discussion_r66334809: see https://github.com/cms-sw/cmssw/compare/CMSSW_8_1_X...diguida:add_back_beam_info_DQM_81X_v1.
Anyhow, I also agree with the proposal by @emeschi in https://github.com/cms-sw/cmssw/pull/14763#discussion_r66354036 to postpone to a later PR.

diguida commented:
@emeschi @mommsen ping :-)
It looks like the bot wants the signature only on the PR comments, not on the code ones.

Some (final) clarifications concerning https://github.com/cms-sw/cmssw/pull/14763#issuecomment-224549386 i.e. the changes in `EventFilter/FEDInterface` and `EventFilter/Utilities`:
* `EventFilter/FEDInterface/interface/FED1024.h`: it is a bit more than "renaming a few fields". I changed the interface of the `TCDSBST` class introducing the `tcdsbst union` instead of a simpler `struct`. The rationale is dual:
 * to make the class `TCDSBST` describe the contents of the BST record in a sensible way, avoiding too many gymnastic on bit masks, with the constraints imposed by padding: this is done by the `TCDSBST::tcdsbst::bst struct`, where some variables were renamed and/or reshuffled for clarity,
 * to provide a hook for the memory address storing the BST record in the TCDS FED raw data, in order to perform more fine grained operations: this is done with the `TCDSBST::tcdsbst::words` array;
* `EventFilter/Utilities/plugins/FRDStreamSource.cc` and `EventFilter/Utilities/src/AuxiliaryMakers.cc`: the changes here are to make both modules compatible with the new TCDS FED API;
* `EventFilter/Utilities/plugins/testGlobalNumbers.cc` add the printouts for the TCDS FED data, and changing coding style according to `clang-format -style=Google`.

Therefore, the modifications in `EventFilter/FEDInterface/interface/FED1024.h` are needed for the purpose of the DQM, as they allow DQM to unpack efficiently the information about LHC configuration and parameters we are requested to provide online.

OTOH, I tend to agree with:

> I would have preferred that we take this on in a more general way.

I think the general solution would be:

1. design a data format entity class suitable for injecting into the EDM format the information coming from BST through TCDS, and, more generally, for all the data packed in the TCDS FED;
2. understand where the product delivered by the data format should be inserted: EventBlock or LuminosityBlock;
3. implement a corresponding unpacker. Currently, there is an unpacker for the TCDS FED raw data, but it provides only the lumi nibble for luminosity measurements and analysis.

This will be beneficial for DQM, too, as the information coming from the TCDS FED (which is our qualified data source) will have stronger correctness guarantee.

My (longish) 2 cents

davidlange6 commented:
@diguida  - while waiting for a daq plus 1, I will comment that the fed 1024 unpacking code needs to move to the tcds unpacker for this PR to go - it should be essentially trivial to add your 15 lines to the code already there and make a new data format (e.g., the proper solution)

diguida commented:
@davidlange6 so you would like me to implement what I proposed as "general solution" in https://github.com/cms-sw/cmssw/pull/14763#issuecomment-225880526:

> 1. design a data format entity class suitable for injecting into the EDM format the information coming from BST through TCDS, and, more generally, for all the data packed in the TCDS FED;
> 2. understand where the product delivered by the data format should be inserted: EventBlock or LuminosityBlock;
> 3. implement a corresponding unpacker. Currently, there is an [unpacker for the TCDS FED raw data](https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/EventFilter/Utilities/plugins/TcdsRawToDigi.cc), but it provides only the lumi nibble for luminosity measurements and analysis.

I can propose an implementation, but then I would like DAQ team (responsible for these pieces of code) to verify it.

Question 1.: in which package is the new entity class is supposed to live? I would say either `DataFormats/FEDRawData` or a brand new `DataFormats/TCDS` package.

davidlange6 commented:
I don't think you need to do "all" as opposed to adding the functionality you already implemented to the proper unpacker.. It looks like DataFormats/Luminosity is a good spot for this (in fact there seems to be an old abandoned class related to this sort of information)

jhegeman commented:
Actually, I don't agree that DataFormats/Luminosity is a good spot. This information is only tangentially related to luminosity. The TCDS event record contains a lot more, general, information. For example: the event number, orbit number, etc. are also taken from this record.

emeschi commented:
I concur with Jeroen. A large fraction of what comes with the TCDS pertains to general information about the event which is used directly by the DAQ code, which is why the “unpacker” was implemented like this.
Whether or not we need a persistent “digi” type is a matter for discussion, but the basic information as to HOW to interpret the TCDS record belongs where it is already and should stay there.
So for David’s requirement it is definitely a -1.
Emilio

> On 15 Jun 2016, at 15:22, Jeroen Hegeman <notifications@github.com> wrote:
> 
> Actually, I don't agree that DataFormats/Luminosity is a good spot. This information is only tangentially related to luminosity. The TCDS event record contains a lot more, general, information. For example: the event number, orbit number, etc. are also taken from this record.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/cms-sw/cmssw/pull/14763#issuecomment-226185239>, or mute the thread <https://github.com/notifications/unsubscribe/AFFwXG8D7LgtUGrXhyY8cLar5SG9wcyKks5qL_yygaJpZM4Itm1j>.
>

emeschi commented:
+1
but not under the conditions described by @davidlange6 <https://github.com/davidlange6>
For the reasons explained in my separate comment.

> On 03 Jun 2016, at 16:50, cmsbuild <notifications@github.com> wrote:
> 
> A new Pull Request was created by @diguida <https://github.com/diguida> (Salvatore Di Guida) for CMSSW_8_1_X.
> 
> It involves the following packages:
> 
> DQM/Integration
> DQMServices/Components
> EventFilter/FEDInterface
> EventFilter/Utilities
> 
> @cvuosalo <https://github.com/cvuosalo>, @emeschi <https://github.com/emeschi>, @dmitrijus <https://github.com/dmitrijus>, @cmsbuild <https://github.com/cmsbuild>, @slava77 <https://github.com/slava77>, @mommsen <https://github.com/mommsen>, @vanbesien <https://github.com/vanbesien>, @davidlange6 <https://github.com/davidlange6> can you please review it and eventually sign? Thanks.
> @Martin-Grunewald <https://github.com/Martin-Grunewald>, @barvic <https://github.com/barvic>, @threus <https://github.com/threus>, @batinkov <https://github.com/batinkov> this is something you requested to watch as well.
> @slava77 <https://github.com/slava77>, @degano <https://github.com/degano>, @smuzaffar <https://github.com/smuzaffar> you are the release manager for this.
> 
> cms-bot commands are list here cms-sw/cmssw#13028 <https://github.com/cms-sw/cmssw/issues/13028>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/cms-sw/cmssw/pull/14763#issuecomment-223600233>, or mute the thread <https://github.com/notifications/unsubscribe/AFFwXKvhcSL-_JTkbLirFzIuWhyK8V_aks5qID87gaJpZM4Itm1j>.
>

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_8_1_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @degano, @smuzaffar

davidlange6 commented:
@emeschi - I don't understand what "unpacker" you mean? By definition DQM does not include unpackers...

diguida commented:
First of all, let me thank @emeschi and @davidlange6 for the discussions we had in the last days.

It seems to me that there are conflicting requirement between DAQ and offline. From one side, DAQ wants to have the possibility to handle efficiently the data produced by TCDS, on the other side offline requires to follow the well-established data flow pattern in CMSSW (FED data are unpacked, encapsulated into entity classes and inserted into dedicated event collections, to be served to DQM/reco clients and so on).
At this point, given that the DQM module is used in production online, works just fine, and is monitored by shift crew in the GUI, I propose the following:

1. we merge the PR as it is in 80X and 81X;
2. I make another PR taking into accounts previous comments, plus a minor change in the appearance of a histogram; alternatively, I can add here all these changes (as a reminder, they are completely unrelated to the PR features), and we relaunch the tests, and approve it quickly;
3. DAQ and offline discuss a proper implementation of an unpacker for the TCDS record (reminder, the [current version](https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/EventFilter/Utilities/plugins/TcdsRawToDigi.cc) only provides the lumi nibble), in particular the BST message: DQM, which runs unpackers in the online sequences, will adapt. As I said before, I can describe in a separate thread (or even here) a possible implementation, as this goes beyond the scope of DQM (and for me, thesis is top priority now).

Let me know.

Commit:
Increase the maximum amount of lumisections we foresee in the x-axis of the histograms populated by DQMServices/Components/src/DQMProvInfo.h

Commit:
Adjust the threshold for testing whether or not collisions are occurring with the beams at the energy allowed for physics production in DQMServices/Components/src/DQMProvInfo.h and DQMServices/Components/src/DQMProvInfo.cc

cmsbuild commented:
Pull request #14763 was updated. @cvuosalo, @emeschi, @dmitrijus, @cmsbuild, @slava77, @mommsen, @vanbesien, @davidlange6 can you please check and sign again.

diguida commented:
The commits https://github.com/cms-sw/cmssw/commit/4c2fa9c4cb8071c485c9a2c864a97b536a0bc77b https://github.com/cms-sw/cmssw/commit/9f8a6f10ceb2de49060f0433aa78d46244dde426 https://github.com/cms-sw/cmssw/commit/f0819572925da85accad4ee0a19b56766d6e778d address comments in https://github.com/cms-sw/cmssw/pull/14763#issuecomment-224717419 https://github.com/cms-sw/cmssw/pull/14763#discussion_r66334307 https://github.com/cms-sw/cmssw/pull/14763#discussion_r66334809

The commit https://github.com/cms-sw/cmssw/commit/3220c1a889585e6639ad695f94f2d2977ed9347f is a workaround for displaying all lumisections of a very long run in a `TH2F` instance.

The commit https://github.com/cms-sw/cmssw/commit/d98e864feeb80cff87266e1c2a3c4c7efca14263 is a fix for handling correctly the beam energy in stable beams, which fluctuates between 6499 and 6500 GeV.

Please start the tests and approve quickly.

emeschi commented:
+1
> On 27 Jun 2016, at 22:40, cmsbuild <notifications@github.com> wrote:
> 
> Pull request #14763 <https://github.com/cms-sw/cmssw/pull/14763> was updated. @cvuosalo <https://github.com/cvuosalo>, @emeschi <https://github.com/emeschi>, @dmitrijus <https://github.com/dmitrijus>, @cmsbuild <https://github.com/cmsbuild>, @slava77 <https://github.com/slava77>, @mommsen <https://github.com/mommsen>, @vanbesien <https://github.com/vanbesien>, @davidlange6 <https://github.com/davidlange6> can you please check and sign again.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/cms-sw/cmssw/pull/14763#issuecomment-228868439>, or mute the thread <https://github.com/notifications/unsubscribe/AFFwXATinlowE_Obc0GJ4VUNIJvO_Abdks5qQDU6gaJpZM4Itm1j>.
>

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: d98e864feeb80cff87266e1c2a3c4c7efca14263
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14763/13710/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14763/13710/summary.html

dmitrijus commented:
+1

cvuosalo commented:
+1

For #14763 d98e864feeb80cff87266e1c2a3c4c7efca14263

Second approval, after minor DQM changes. New Jenkins tests are still OK.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_8_1_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @degano, @smuzaffar

dmitrijus commented:
@davidlange6 ping!
